src/opencl_ansible_fmt_plug.c

```diff
@@ -0,0 +1,343 @@
+
+		// Initialize openCL tuning (library) for this format.
+		opencl_init_auto_setup(SEED, HASH_LOOPS, split_events, warn,
+		                       2, self, create_clobj, release_clobj,
+		                       sizeof(state_t), 0, db);
+
+		// Auto tune execution from shared/included code.
+		autotune_run(self, ITERATIONS, 0,
+		             (cpu(device_info[gpu_id]) ?
+		              1000000000 : 10000000000ULL));
+	}
+}
+
+static void done(void)
+{
+	if (autotuned) {
+		release_clobj();
+		HANDLE_CLERROR(clReleaseKernel(crypt_kernel), "Release kernel 1");
+		HANDLE_CLERROR(clReleaseKernel(split_kernel), "Release kernel 2");
+		HANDLE_CLERROR(clReleaseKernel(final_kernel), "Release kernel 3");
+		HANDLE_CLERROR(clReleaseProgram(program[gpu_id]),
+		               "Release Program");
+
+		autotuned--;
+	}
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt*)salt;
+
+	memcpy(host_salt->salt, cur_salt->salt, cur_salt->salt_length);
+	host_salt->length = cur_salt->salt_length;
+	host_salt->rounds = cur_salt->iterations;
+	host_salt->outlen = 64;
+
+	HANDLE_CLERROR(clEnqueueWriteBuffer(queue[gpu_id], mem_salt,
+		CL_FALSE, 0, sizeof(salt_t), host_salt, 0, NULL, NULL),
+	    "Copy salt to gpu");
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	int i = 0, j = 0;
+	const int count = *pcount;
+	int index;
+	int loops = (host_salt->rounds + HASH_LOOPS - 1) / HASH_LOOPS;
+	size_t *lws = local_work_size ? &local_work_size : NULL;
+
+	global_work_size = GET_MULTIPLE_OR_BIGGER(count, local_work_size);
+
+	// Copy data to gpu
+	BENCH_CLERROR(clEnqueueWriteBuffer(queue[gpu_id], mem_in,
+		CL_FALSE, 0, global_work_size * sizeof(pass_t), host_pass, 0,
+		NULL, multi_profilingEvent[0]), "Copy data to gpu");
+
+	// Run kernel
+	BENCH_CLERROR(clEnqueueNDRangeKernel(queue[gpu_id], crypt_kernel,
+		1, NULL, &global_work_size, lws, 0, NULL, multi_profilingEvent[1]), "Run kernel");
+	for (j = 0; j < (ocl_autotune_running ? 1 : (host_salt->outlen + 31) / 32); j++) {
+		for (i = 0; i < (ocl_autotune_running ? 1 : loops); i++) {
+			BENCH_CLERROR(clEnqueueNDRangeKernel(queue[gpu_id], split_kernel,
+				1, NULL, &global_work_size, lws, 0, NULL,
+				multi_profilingEvent[2]), "Run split kernel");
+			BENCH_CLERROR(clFinish(queue[gpu_id]), "Error running loop kernel");
+			opencl_process_event();
+		}
+
+		BENCH_CLERROR(clEnqueueNDRangeKernel(queue[gpu_id], final_kernel,
+			1, NULL, &global_work_size, lws, 0, NULL,
+			multi_profilingEvent[3]), "Run final kernel");
+	}
+
+	// Read the result back
+	BENCH_CLERROR(clEnqueueReadBuffer(queue[gpu_id], mem_out,
+		CL_TRUE, 0, global_work_size * sizeof(crack_t), host_crack, 0,
+		NULL, multi_profilingEvent[4]), "Copy result back");
+
+	if (!ocl_autotune_running) {
+		for (index = 0; index < count; index++) {
+			JTR_hmac_sha256(((unsigned char*)host_crack[index].hash) + 32, 32, cur_salt->blob, cur_salt->bloblen, (unsigned char*)crypt_out[index], 16);
+		}
+	}
```

Comment:
The HMAC should obviously be made on GPU as well. If we don't have a prefab file for it, a HMAC is really basic stuff you can do "manually" with the SHA-256 ctx code.

Reply:
I am running out of `JtR time` today. I will do it in a future PR. Thanks!

Reply:
I'm merging this for now - I'll have a look at "skip bytes" today! Perhaps I have time to fix the HMAC as well.